### PR TITLE
feat(47983): Negative tuple index access should not be allowed

### LINKED ISSUE #39293: Negative tuple indexer should not be valid
**TypeScript Version:**  4.0.0-dev.20200625

**Search Terms:**
negative tuple indexer

**Code**

```ts
type t1 = [string, number][0]; // string
type t2 = [string, number][1]; // number
type t3 = [string, number][-1]; // (string | number)  ???

type u<t extends any[], u extends number> = t[u];

type u1 = u<[string, number], 0>; // string
type u2 = u<[string, number], 1>; // number
type u3 = u<[string, number], 2>; // undefined
type u4 = u<[string, number], 1.5>; // undefined
type u5 = u<[string, number], -0.5>; // (string | number)  ???
```

**Expected behavior:**
that `type t3 = [string, number][-1];` emits a compile-time error "Tuple type '[string, number]' of length '2' has no element at index '-1'." and `u5` is `undefined`.

**Actual behavior:**
`t3` and `u5` is a union of the types contained in the tuple.

**Playground Link:** [playground link](https://www.typescriptlang.org/play/?ts=4.0.0-dev.20200625#code/C4TwDgpgBMCMUF4oG0DOwBOBLAdgcwBoocBXAWwCMIMBdZABhoG4oB6VqdbfAWAChQkGACZEKLrkLFyVWsljM2HUpWr9B0YAGYxaTJKIrZdALQKW7KAAoJ+KAB9pqjAEooUAPxf+68NBIAPMBQEAAewBA4ACaoUACGOCDINEQkIeGRMU6yAHxiwMgkzD4CflAk8EiBetxSRtQpUPQ5Fhy2eL5CJKJVATUG2Q1EsC1Kgxid-jq9-fiGMkNQwqOWJNEQAGa4EFGT5QAsYtXt886NsAB0AKwrHGtRm9u7pV1XR30n440m9Ne31u0HOM3J5vHwgA)

### LINKED ISSUE #47983: Negative tuple index access should not be allowed
# Bug Report

### 🔎 Search Terms
`tuple,` `negative`

Found #39293 with the same remarks but regarding types, not values.

### 🕗 Version & Regression Information
Observed in 4.5.5, 4.5.4 (Playground) and 3.9.7 (Playground)

### ⏯ Playground Link
There: [Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4&q=141#code/MYewdgzgLgBAtgTwCoFcAOaA2BTAXDAbRjBTgCNsAnAGmNIspgF0YBeQmABls+YFgAUIID0wmAHEQIACa00AQwgRsEQaEiwQAazbxk6LNgIBGJoJFjJMmFCm0AZvICWmCDEUxsADzTZgUbGkYAHcnKAALGAAiVEMbBF8YAHICEnIqWjSGJiSYEHsYHDAAcwjkgCZc8I8wEE8cOGwwWHlYJzBpb2SAZiSAOhsIAApygBYATm6ASii1cGg8rXLdRFjDAm6WGHMBURgAORBYYqlZGAhwkBRMIKpKEEYrqDmNPJA0N3ZVgxwCAFpTIIgA)

### 💻 Code
```ts
const myTupple: [ number, number ] = [ 0, 0 ]

// Good, passes
const ok = myTupple[1]

// Good too, fails as expected with "Tuple type '[number, number]' of length '2' has no element at index '3'. ts(2493)"
const ok2 = myTupple[3]  

// Not good, should error out
const oops = myTupple[-1]

```

### 🙁 Actual behavior
Passes.

### 🙂 Expected behavior
Should fail (I guess).

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #47983
Fixes #39293

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2356,6 +2356,10 @@
         "category": "Error",
         "code": 2513
     },
+    "Cannot use a negative integer for indexing on '{0}'.": {
```
Comment (jakebailey) [inline src/compiler/diagnosticMessages.json:2359]: The current wording for this sort of bounds check is:

`Tuple type '{0}' of length '{1}' has no element at index '{2}'.`

So, maybe something like:

- `Tuple type cannot be indexed negatively`

Or just let it say "Tuple type ... of length ... has no element at index '-1'."?

@DanielRosenwasser

Comment (DanielRosenwasser) [reply src/compiler/diagnosticMessages.json:2359]: > A tuple type cannot be indexed with a negative value.

**Review by jakebailey — APPROVED** [review]

**Comment by jakebailey** [comment]
I think this actually fixes #39293 as well, if you copy in the test for it.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -15581,22 +15581,29 @@ namespace ts {
...
                 }
-                if (everyType(objectType, isTupleType) && isNumericLiteralName(propName) && +propName >= 0) {
+                if (everyType(objectType, isTupleType) && isNumericLiteralName(propName)) {
+                    const index = +propName;
                     if (accessNode && everyType(objectType, t => !(t as TupleTypeReference).target.hasRestElement) && !(accessFlags & AccessFlags.NoTupleBoundsCheck)) {
                         const indexNode = getIndexNodeForAccessExpression(accessNode);
                         if (isTupleType(objectType)) {
+                            if (index < 0) {
+                                error(indexNode, Diagnostics.A_tuple_type_cannot_be_indexed_with_a_negative_value);
+                                return undefinedType;
```
Comment (jakebailey) [inline src/compiler/checker.ts:15591]: Actually, should this be `errorType`?

**Review by jakebailey — CHANGES_REQUESTED** [review]
```
    type t1 = [string, number][0]; // string
    type t2 = [string, number][1]; // number
    type t3 = [string, number][-1]; // (string | number)  ???
                               ~~
!!! error TS2514: A tuple type cannot be indexed with a negative value. 
```

Here's the output if I add the test in a local checkout, which is good, but I am wondering if it's better to do an error type instead? I suppose you chose undefined because negative can only ever be out of bounds, but we pick a different choice for values that are out of bounds the other way, which is curious.

File: src/compiler/checker.ts (hunk 2 above)
Comment (jakebailey) [reply src/compiler/checker.ts:15591]: No, I guess not: 
image

**Comment by jakebailey** [comment]
Nevermind about errorType, `undefined` seems OK based on the existing positive out of range behavior, but I would add a test for that other issue.

**Review by jakebailey — APPROVED** [review]
LGTM, but looking at #39293 it seemed like we weren't sure. I think that it's fine, personally. @DanielRosenwasser?

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Review by sandersn — APPROVED** [review]

**Pull request merged**